service_discovery/core/src/registry_impl.rs

```diff
@@ -0,0 +1,457 @@
+        // This block controls the lifetime of the lock.
+        {
+            let mut lock: RwLockWriteGuard<HashMap<ServiceIdentifiers, ServiceMetadata>> =
+                self.registry_map.write();
+            match lock.get(&service_identifiers) {
+                Some(_) => {
+                    lock.insert(service_identifiers.clone(), service_to_register.clone());
+                    registration_status = RegistrationStatus::Updated;
+                    info!("Updated the service entry in Chariott service registry; overwrote previous entry: {:?}", service_to_register);
+                }
+                None => {
+                    lock.insert(service_identifiers.clone(), service_to_register.clone());
+                    registration_status = RegistrationStatus::NewlyRegistered;
+                    info!(
+                        "Registered new service in Chariott service registry: {:?}",
+                        service_to_register
+                    );
+                }
+            };
+        }
+        let register_response =
+            RegisterResponse { registration_status: registration_status as i32 };
+        Ok(Response::new(register_response))
+    }
+
+    /// Unregisters a service by removing it from the registry.
+    ///
+    /// This function registers a service based on a [`UnregisterRequest`]. Returns a
+    /// [`UnregisterResponse`].
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - A [`UnregisterRequest`] wrapped by a [`tonic::Request`].
+    async fn unregister(
+        &self,
+        request: Request<UnregisterRequest>,
+    ) -> Result<Response<UnregisterResponse>, Status> {
+        let request_inner = request.into_inner();
+        let service_identifiers = ServiceIdentifiers {
+            namespace: request_inner.namespace,
+            name: request_inner.name,
+            version: request_inner.version,
+        };
+        info!("Received an unregister request for: {:?}", service_identifiers);
+
+        // This block controls the lifetime of the lock.
+        {
+            let mut lock: RwLockWriteGuard<HashMap<ServiceIdentifiers, ServiceMetadata>> =
+                self.registry_map.write();
+            match lock.remove(&service_identifiers) {
+                Some(removed_service) => {
+                    info!(
+                        "Successfully removed service entry in Chariott service registry: {:?}",
+                        removed_service
+                    );
+                }
+                None => {
+                    let not_found_message = format!(
+                        "Unable to remove service from registry: {:?}",
+                        service_identifiers
+                    );
+                    warn!(not_found_message); //namespace: {0}, name: {1}, version: {2}")
+                    return Err(Status::not_found(not_found_message));
+                }
+            };
+        }
+        Ok(Response::new(UnregisterResponse {}))
```

Comment:
I would move this to the end of the `Some` arm of the match and omit the `return` in front of `Err`. So it would have this structure:

```
match lock.remove(...) {
    Some(...) => {
        info!(...);
        Ok(Response::new(UnregisterResponse {}))
    }
    None => {
        warn!(...);
        Err(Status::not_found(not_found_message))
    }
}
```